### ISSUE TITLE
PathTooLongException when trying to add a NuGet package to empty ASP.NET Core web application

### ISSUE THREAD

**Issue by balachir** [NuGet/Home#4391]
## Steps to reproduce

1. Using VS 2017 RC.3 (d15rel.26123.00)
2. New ASP.NET Core 1.0 web app > Empty template
    - created my project in this path: ```C:\Repos\balatest01\ProjectKTestProjects\Dev15_RC3\final\WebApp_Anc10_core_Empty```
3. Hit F5, works
4. Open Manage NuGet packages dialog
5. Search for "asp.netcore.staticfiles" > choose the default 1.1.0 version and hit Install

## Actual behavior
Error list comes up with this error

![image](https://cloud.githubusercontent.com/assets/8246794/22262748/f91670e8-e226-11e6-98b5-f4cc76dd57ee.png)

Here are the contents of my Output Window (Package Manager pane)
```
Restoring packages for C:\Repos\balatest01\ProjectKTestProjects\Dev15_RC3\final\WebApp_Anc10_core_Empty\WebApp_Anc10_core_Empty\WebApp_Anc10_core_Empty.csproj...
  GET https://api.nuget.org/v3-flatcontainer/runtime.debian.8-x64.runtime.native.system.security.cryptography.openssl/index.json
  GET https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.openssl/index.json
  GET https://api.nuget.org/v3-flatcontainer/runtime.rhel.7-x64.runtime.native.system.security.cryptography.openssl/index.json
  GET https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.14.04-x64.runtime.native.system.security.cryptography.openssl/index.json
  GET https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.04-x64.runtime.native.system.security.cryptography.openssl/index.json
  GET https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.10-x64.runtime.native.system.security.cryptography.openssl/index.json
  OK https://api.nuget.org/v3-flatcontainer/runtime.debian.8-x64.runtime.native.system.security.cryptography.openssl/index.json 252ms
  GET https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.apple/index.json
  GET https://api.nuget.org/v3-flatcontainer/runtime.debian.8-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.debian.8-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg
  OK https://api.nuget.org/v3-flatcontainer/runtime.rhel.7-x64.runtime.native.system.security.cryptography.openssl/index.json 240ms
  GET https://api.nuget.org/v3-flatcontainer/runtime.rhel.7-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.rhel.7-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg
  OK https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.14.04-x64.runtime.native.system.security.cryptography.openssl/index.json 247ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.openssl/index.json 252ms
  GET https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.14.04-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.ubuntu.14.04-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg
  GET https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg
  OK https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.04-x64.runtime.native.system.security.cryptography.openssl/index.json 288ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.10-x64.runtime.native.system.security.cryptography.openssl/index.json 306ms
  GET https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.04-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.ubuntu.16.04-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg
  GET https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.10-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.ubuntu.16.10-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg
  OK https://api.nuget.org/v3-flatcontainer/runtime.debian.8-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.debian.8-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg 251ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.rhel.7-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.rhel.7-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg 232ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.apple/index.json 257ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.14.04-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.ubuntu.14.04-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg 233ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg 235ms
  GET https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.apple/4.3.0/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.apple.4.3.0.nupkg
  OK https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.04-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.ubuntu.16.04-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg 236ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.ubuntu.16.10-x64.runtime.native.system.security.cryptography.openssl/4.3.0/runtime.ubuntu.16.10-x64.runtime.native.system.security.cryptography.openssl.4.3.0.nupkg 236ms
  OK https://api.nuget.org/v3-flatcontainer/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.apple/4.3.0/runtime.osx.10.10-x64.runtime.native.system.security.cryptography.apple.4.3.0.nupkg 1075ms
Installing runtime.osx.10.10-x64.runtime.native.System.Security.Cryptography.Apple 4.3.0.
Installing runtime.ubuntu.16.10-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing runtime.ubuntu.16.04-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing runtime.osx.10.10-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing runtime.ubuntu.14.04-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing runtime.debian.8-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing runtime.rhel.7-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing runtime.opensuse.42.1-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing Microsoft.AspNetCore.StaticFiles 1.1.0.
Installing Microsoft.AspNetCore.Hosting.Abstractions 1.1.0.
Installing runtime.opensuse.13.2-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing Microsoft.AspNetCore.Http.Extensions 1.1.0.
Installing Microsoft.Extensions.FileProviders.Abstractions 1.1.0.
Installing Microsoft.Extensions.Logging.Abstractions 1.1.0.
Installing Microsoft.Extensions.WebEncoders 1.1.0.
Installing runtime.fedora.24-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing NETStandard.Library 1.6.1.
Installing Microsoft.AspNetCore.Hosting.Server.Abstractions 1.1.0.
Installing runtime.fedora.23-x64.runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing System.Security.Principal 4.3.0.
Installing Microsoft.AspNetCore.Http.Abstractions 1.1.0.
Installing System.Security.Claims 4.3.0.
Installing System.Net.WebSockets 4.3.0.
Installing Microsoft.Extensions.Configuration.Abstractions 1.1.0.
Installing runtime.native.System.Security.Cryptography.Apple 4.3.0.
Installing Microsoft.Extensions.DependencyInjection.Abstractions 1.1.0.
Installing System.Security.Cryptography.Csp 4.3.0.
Installing Microsoft.Net.Http.Headers 1.1.0.
Installing System.Buffers 4.3.0.
Installing Microsoft.Extensions.Primitives 1.1.0.
Installing Microsoft.Extensions.Options 1.1.0.
Installing System.Text.Encodings.Web 4.3.0.
Installing Microsoft.NETCore.Platforms 1.1.0.
Installing Microsoft.Win32.Primitives 4.3.0.
Installing System.AppContext 4.3.0.
Installing System.Collections 4.3.0.
Installing System.Collections.Concurrent 4.3.0.
Installing System.Console 4.3.0.
Installing System.Diagnostics.Debug 4.3.0.
Installing System.Diagnostics.Tools 4.3.0.
Installing System.Diagnostics.Tracing 4.3.0.
Installing System.Globalization 4.3.0.
Installing System.Globalization.Calendars 4.3.0.
Installing System.Security.Cryptography.Cng 4.3.0.
Installing System.IO 4.3.0.
Installing System.IO.Compression 4.3.0.
Installing System.IO.Compression.ZipFile 4.3.0.
Installing System.IO.FileSystem 4.3.0.
Installing System.IO.FileSystem.Primitives 4.3.0.
Installing System.Linq 4.3.0.
Installing System.Linq.Expressions 4.3.0.
Installing System.Net.Primitives 4.3.0.
Installing System.Net.Sockets 4.3.0.
Installing System.ObjectModel 4.3.0.
Installing System.Reflection 4.3.0.
Installing System.Reflection.Extensions 4.3.0.
Installing System.Reflection.Primitives 4.3.0.
Installing System.Resources.ResourceManager 4.3.0.
Installing System.Runtime 4.3.0.
Installing System.Runtime.Extensions 4.3.0.
Installing System.Runtime.Handles 4.3.0.
Installing System.Runtime.InteropServices 4.3.0.
Installing System.Runtime.InteropServices.RuntimeInformation 4.3.0.
Installing System.Runtime.Numerics 4.3.0.
Installing System.Security.Cryptography.Primitives 4.3.0.
Installing System.Text.Encoding 4.3.0.
Installing System.Text.Encoding.Extensions 4.3.0.
Installing System.Text.RegularExpressions 4.3.0.
Installing System.Threading 4.3.0.
Installing System.Threading.Tasks 4.3.0.
Installing System.Threading.Timer 4.3.0.
Installing System.Xml.ReaderWriter 4.3.0.
Installing System.Xml.XDocument 4.3.0.
Installing System.Net.Http 4.3.0.
Installing System.Security.Cryptography.Encoding 4.3.0.
Installing System.Security.Cryptography.X509Certificates 4.3.0.
Installing System.Security.Cryptography.Algorithms 4.3.0.
Installing Microsoft.AspNetCore.Http.Features 1.1.0.
Installing System.Globalization.Extensions 4.3.0.
Installing System.Reflection.TypeExtensions 4.3.0.
Installing System.ComponentModel 4.3.0.
Installing System.Diagnostics.Contracts 4.3.0.
Installing System.Runtime.CompilerServices.Unsafe 4.3.0.
Installing Microsoft.NETCore.Targets 1.1.0.
Installing runtime.native.System 4.3.0.
Installing runtime.native.System.IO.Compression 4.3.0.
Installing System.Reflection.Emit 4.3.0.
Installing System.Reflection.Emit.ILGeneration 4.3.0.
Installing runtime.native.System.Security.Cryptography.OpenSsl 4.3.0.
Installing System.Reflection.Emit.Lightweight 4.3.0.
Installing System.Threading.Tasks.Extensions 4.3.0.
Installing runtime.native.System.Net.Http 4.3.0.
Installing System.Diagnostics.DiagnosticSource 4.3.0.
Installing System.Security.Cryptography.OpenSsl 4.3.0.
Installing System.Private.Uri 4.3.0.
Installing System.Security.Principal.Windows 4.3.0.
Installing System.Net.NameResolution 4.3.0.
Installing System.Threading.Overlapped 4.3.0.
Installing runtime.win.Microsoft.Win32.Primitives 4.3.0.
Installing runtime.any.System.Threading.Timer 4.3.0.
Installing runtime.any.System.Threading.Tasks 4.3.0.
Installing runtime.any.System.Text.Encoding.Extensions 4.3.0.
Installing runtime.any.System.Text.Encoding 4.3.0.
Installing runtime.any.System.Collections 4.3.0.
Installing runtime.any.System.Runtime.InteropServices 4.3.0.
Installing runtime.win.System.Console 4.3.0.
Installing runtime.any.System.Runtime.Handles 4.3.0.
Installing runtime.win.System.Diagnostics.Debug 4.3.0.
Installing runtime.win.System.Runtime.Extensions 4.3.0.
Installing runtime.any.System.Runtime 4.3.0.
Installing runtime.any.System.Diagnostics.Tools 4.3.0.
Installing runtime.any.System.Diagnostics.Tracing 4.3.0.
Installing runtime.any.System.Globalization 4.3.0.
Installing runtime.any.System.Resources.ResourceManager 4.3.0.
Installing runtime.any.System.Reflection.Primitives 4.3.0.
Installing runtime.any.System.Globalization.Calendars 4.3.0.
Installing runtime.any.System.IO 4.3.0.
Installing runtime.win.System.IO.FileSystem 4.3.0.
Installing runtime.win.System.Net.Primitives 4.3.0.
Installing runtime.win.System.Net.Sockets 4.3.0.
Installing runtime.any.System.Reflection.Extensions 4.3.0.
Installing runtime.any.System.Reflection 4.3.0.
System.IO.PathTooLongException: The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.
   at System.IO.PathHelper.GetFullPathName()
   at System.IO.Path.LegacyNormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
   at System.IO.Path.NormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
   at System.IO.Path.GetFullPathInternal(String path)
   at System.IO.FileInfo.Init(String fileName, Boolean checkHost)
   at System.IO.FileInfo..ctor(String fileName)
   at NuGet.Protocol.LocalFolderUtility.GetPackagesConfigFolderPackage(String root, PackageIdentity identity, ILogger log)
   at NuGet.Protocol.FindLocalPackagesResourcePackagesConfig.GetPackage(PackageIdentity identity, ILogger logger, CancellationToken token)
   at NuGet.Protocol.LocalPackageMetadataResource.<>c__DisplayClass3_0.<GetMetadataAsync>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.UI.UIActionEngine.<GetPackageMetadataAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.UI.TaskCombinators.<>c__DisplayClass1_0`2.<<ThrottledAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.UI.TaskCombinators.<ThrottledAsync>d__1`2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.UI.UIActionEngine.<GetPackageMetadataAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.UI.UIActionEngine.<CheckLicenseAcceptanceAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.PackageManagement.UI.UIActionEngine.<PerformActionImplAsync>d__7.MoveNext()
Time Elapsed: 00:00:36.0043048
========== Finished ==========


```

## Expected  behavior
Package restore to succeed

**Comment by balachir** [comment]
Reproduces when I created the ASP.NET Core 1.0 project in the default path as well
```
C:\Users\balach\Documents\visual studio 2017\Projects\WebApp1_Anc10_core_Empty
```

Does not repro if I create an ASP.NET Core 1.1 project and try to add the Microsoft.AspNetCore.StaticFiles 1.1.0 package.

**Comment by emgarten** [comment]
It seems like the issue here is the length of the package ids, those should be shortened.

To improve this in NuGet we would need to completely change the output folder structure which would be a breaking change.

**Comment by dsplaisted** [comment]
I hit this issue also.  The method names on the stack make it seem like it might be trying to calculate the path for a packages.config project:

>    at NuGet.Protocol.LocalFolderUtility.GetPackagesConfigFolderPackage(String root, PackageIdentity identity, ILogger log)
   at NuGet.Protocol.FindLocalPackagesResourcePackagesConfig.GetPackage(PackageIdentity identity, ILogger logger, CancellationToken token)
   at NuGet.Protocol.LocalPackageMetadataResource.<>c__DisplayClass3_0.<GetMetadataAsync>b__0()

However, this is with an ASP.NET Core project that uses `PackageReference`, so the packages.config folder shouldn't matter.  Or is this not what it is doing?

**Comment by dsplaisted** [comment]
I tried the same thing with a project under a shorter path and I did not hit this error.  So it does seem like what is happening is that NuGet is applying package.config path length limitations based on the project path to `PackageReference` projects.

**Comment by emgarten** [comment]
> so the packages.config folder shouldn't matter. Or is this not what it is doing?

This might be due to the UI trying to reload details on the package. In that case I think the exception could be ignored on the nuget side.

Are you seeing the same issue when using powershell with ``install-package``?

**Comment by rrelyea** [comment]
@emgarten  - how common will this be hit. moved back to rtm, cause it feels pretty bad. thoughts?

**Comment by dsplaisted** [comment]
@emgarten Rather than ignoring the exception, I would expect it to never even go down that code path if the project is using `PackageReference`.

@rrelyea I think this will be very common.  My username is only 8 characters long, and I hit this with a project named "WebApiHostPolicyRepro" in the default VS project location: `C:\Users\daplaist\Documents\Visual Studio 2017\Projects\WebApiHostPolicyRepro\WebApiHostPolicyRepro`

**Comment by emgarten** [comment]
>  I would expect it to never even go down that code path if the project is using PackageReference

There are other scenarios such as managing packages at the solution level where the same package is installed into a packages.config, project.json, and PackageReference project. In short, the UI is just trying to be efficient by looking around for a nuspec first before going online to read all sources for the same info.

> how common will this be hit. moved back to rtm, cause it feels pretty bad. thoughts?

We should investigate this for RTM. We are likely just doing a file exists check on a long path when probing and it is throwing, we should wrap these calls and handle this better.

**Comment by rrelyea** [comment]
Unless this is common, we'll address in 4.0.1.

**Comment by dsplaisted** [comment]
@rrelyea To me, it seems like this could be common.

Below are the repro steps for me.  I'm not sure exactly which package is triggering this but it seems like manually installing or upgrading to the ASP.NET 1.1 packages will often hit this bug, depending on the path length of where your project is stored.

- File \ New Project \ ASP.NET Core Web Application (.NET Core)
- Choose a name for the project that is 20 characters long (It may repro with shorter names too, I haven't determined exactly how long the project name needs to be, and it will also depend on the length of your username)
- Select the Web API template for ASP.NET Core 1.0 and click OK
- Go to Manage NuGet Packages, go to the updates tab, and install all updates

EXPECTED: Updates install successfully
ACTUAL: Updates fail with PathTooLongException

**Comment by rohit21agrawal** [comment]
@rrelyea i think this needs to be fixed in 4.0 RTM

**Comment by rrelyea** [comment]
Turns out this seems very common.
